#! /bin/sh -e
## skinny.dpatch by Kilian Krause <kk@verfaction.de>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: fixes skinny bug http://bugs.digium.com/bug_view_page.php?bug_id=0003496

if [ $# -lt 1 ]; then
    echo "`basename $0`: script expects -patch|-unpatch as argument" >&2
    exit 1
fi

[ -f debian/patches/00patch-opts ] && . debian/patches/00patch-opts
patch_opts="${patch_opts:--f --no-backup-if-mismatch} ${2:+-d $2}"

case "$1" in
    -patch) patch -p1 ${patch_opts} < $0;;
    -unpatch) patch -R -p1 ${patch_opts} < $0;;
    *)
        echo "`basename $0`: script expects -patch|-unpatch as argument" >&2
        exit 1;;
esac

exit 0

@DPATCH@
diff -Nru asterisk/channels/chan_skinny.c asterisk.new/channels/chan_skinny.c
--- asterisk/channels/chan_skinny.c	2005-01-23 03:35:40.000000000 +0000
+++ asterisk.new/channels/chan_skinny.c	2005-02-01 16:06:14.517006556 +0000
@@ -1576,16 +1576,20 @@
 static struct ast_frame *skinny_rtp_read(struct skinny_subchannel *sub)
 {
 	/* Retrieve audio/etc from channel.  Assumes sub->lock is already held. */
-	struct ast_frame *f;
-	f = ast_rtp_read(sub->rtp);
-	if (sub->owner) {
-		/* We already hold the channel lock */
-		if (f->frametype == AST_FRAME_VOICE) {
-			if (f->subclass != sub->owner->nativeformats) {
-				ast_log(LOG_DEBUG, "Oooh, format changed to %d\n", f->subclass);
-				sub->owner->nativeformats = f->subclass;
-				ast_set_read_format(sub->owner, sub->owner->readformat);
-				ast_set_write_format(sub->owner, sub->owner->writeformat);
+	static struct ast_frame null_frame = { AST_FRAME_NULL, };
+	struct ast_frame *f = &null_frame;
+	if (sub->rtp)
+	{
+		f = ast_rtp_read(sub->rtp);
+		if (sub->owner) {
+			/* We already hold the channel lock */
+			if (f->frametype == AST_FRAME_VOICE) {
+				if (f->subclass != sub->owner->nativeformats) {
+					ast_log(LOG_DEBUG, "Oooh, format changed to %d\n", f->subclass);
+					sub->owner->nativeformats = f->subclass;
+					ast_set_read_format(sub->owner, sub->owner->readformat);
+					ast_set_write_format(sub->owner, sub->owner->writeformat);
+				}
 			}
 		}
 	}
